File: haskell/json.hs
{-# OPTIONS_GHC -Wall #-}

import           Text.ParserCombinators.Parsec

data JsonValue
  = String String
  | Number Int
  | Array [JsonValue]
  | Bool Bool
  | Null
  | Object [(String, JsonValue)]
  deriving (Show, Eq)

parseJson :: String -> Maybe JsonValue
parseJson str =
  case parse json "" str of
    Left _  -> Nothing
    Right v -> Just v

json :: Parser JsonValue
json = do
  whiteSpace
  value <- jsonValue
  eof
  return value

jsonValue :: Parser JsonValue
jsonValue = do
  whiteSpace
  value <-
    try (string "null" >> return Null)
      <|> try (string "true" >> return (Bool True))
      <|> try (string "false" >> return (Bool False))
      <|> try jsonString
      <|> try jsonNumber
      <|> try jsonArray
      <|> try jsonObject
  whiteSpace
  return value

jsonString :: Parser JsonValue
jsonString = do
  _ <- char '"'
  str <- many (noneOf "\"")
  _ <- char '"'
  return $ String str

jsonNumber :: Parser JsonValue
jsonNumber = do
  str <- many1 digit
  return $ Number (read str)

jsonArray :: Parser JsonValue
jsonArray = do
  _ <- char '['
  whiteSpace
  values <- sepBy jsonValue (char ',')
  whiteSpace
  _ <- char ']'
  return $ Array values

jsonObject :: Parser JsonValue
jsonObject = do
  _ <- char '{'
  whiteSpace
  pairs <- sepBy jsonPair (char ',')
  whiteSpace
  _ <- char '}'
  return $ Object pairs

jsonPair :: Parser (String, JsonValue)
jsonPair = do
  whiteSpace
  key <- jsonString
  whiteSpace
  _ <- char ':'
  whiteSpace
  value <- jsonValue
  whiteSpace
  return (getString key, value)

getString :: JsonValue -> String
getString (String str) = str
getString _            = error "not a string"

whiteSpace :: Parser ()
whiteSpace = skipMany (oneOf " \t\n")

File: json.pl
parse(String, Data) :-
    atom_chars(String, Chars),
    phrase(json(Data), Chars).

json(String) --> json_string(String).
json(Number) --> json_number(Number).
json(Bool) --> json_bool(Bool).
json(Null) --> json_null(Null).
json(Array) --> json_array(Array).
json(Object) --> json_object(Object).

json_string(String) --> ['"'], json_chars(Chars), ['"'], { atom_chars(String, Chars) }.

json_chars([]) --> [].
json_chars([Char|Chars]) --> [Char], { Char \= '"' }, json_chars(Chars).

json_number(Number) --> json_digits(Digits), { number_chars(Number, Digits) }.
json_digits([Digit|Digits]) --> json_digit(Digit), json_digits(Digits).
json_digits([Digit]) --> json_digit(Digit).
json_digit(Digit) --> [Digit], { code_type(Digit, digit) }.

json_bool(true) --> ['t','r','u','e'].
json_bool(false) --> ['f','a','l','s','e'].

json_null(null) --> ['n','u','l','l'].

json_array(Array) --> ['['], json_values(Array), [']'].

json_values([Value|Values]) --> json(Value), [','], json_values(Values).
json_values([Value]) --> json(Value).

json_object(Object) --> ['{'], json_pairs(Object), ['}'].
json_pairs([Pair|Pairs]) --> json_pair(Pair), [','], json_pairs(Pairs).
json_pairs([Pair]) --> json_pair(Pair).
json_pair(Key=Value) --> json_string(Key), [':'], json(Value).

